### PR TITLE
[AC-2509] add member access component

### PR DESCRIPTION
**Pull request by aj-rosado**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/AC-2509
## 📔 Objective
This focus on the visual components and layouts and not on the service layer of the Member Access Report page.

Add the MemberAccessReportComponent on the `bitwarden_license`, it is using static data for now.
Also added navigation to the component on the `organization-reporting-routing`
Created `member-access-report.service.ts` that has a temporary method to return static data

### Light Theme
image

### Dark Theme
image

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: bitwarden_license/bit-web/src/app/tools/reports/member-access-report/member-access-report.component.html [hunk 1]
```diff
@@ -0,0 +1,52 @@
+<ng-container>
+  <app-header>
```
Comment (djsmith85) [inline bitwarden_license/bit-web/src/app/tools/reports/member-access-report/member-access-report.component.html:2]: ❓  Does this need to be wrapped within an ng-container?

Other report pages are structured as:

```
<app-header></app-header>

<bit-container>
 table content here...
</bit-container>
```

**Review by djsmith85 — CHANGES_REQUESTED** [review]
Report looks good just some small items

:thought_balloon: Maybe if we made the member-access-report standalone, it would simplfy and centralize the routing and guarding as we wouldn't need a specific reports-reporting.module for import.

File: apps/web/src/app/admin-console/organizations/reporting/organization-reporting-routing.module.ts [hunk 2]
```diff
@@ -79,6 +79,13 @@ const routes: Routes = [
             },
             canActivate: [IsPaidOrgGuard],
           },
+          {
+            path: "",
+            loadChildren: () =>
+              import(
+                "../../../../../../../bitwarden_license/bit-web/src/app/tools/reports/reports.module"
```
[the remark this answers is not in the document]
Comment (eliykat) [reply apps/web/src/app/admin-console/organizations/reporting/organization-reporting-routing.module.ts:86]: To expand slightly, see [OrganizationRoutingModule](https://github.com/bitwarden/clients/blob/main/apps/web/src/app/admin-console/organizations/organization-routing.module.ts#L1) (OSS) and [OrganizationsRoutingModule](https://github.com/bitwarden/clients/blob/main/bitwarden_license/bit-web/src/app/admin-console/organizations/organizations-routing.module.ts#L1) (BL). They're both standard routing modules, but it shows how you can declare a separate routing module in the Bitwarden-licensed code and have it "overlap" with your standard routing module (i.e. even though their paths partially overlap, Angular still resolves them in order).

Also note the [order that they are registered in](https://github.com/bitwarden/clients/blob/main/bitwarden_license/bit-web/src/app/app.module.ts#L41), which might be a gotcha.

File: bitwarden_license/bit-web/src/app/tools/reports/reports-routing.module.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import { MemberAccessReportComponent } from "./member-access-report/member-access-report.component";
+
+const routes: Routes = [
+  {
+    path: "member-access-report",
+    component: MemberAccessReportComponent,
+    data: {
+      titleId: "memberAccessReport",
+    },
+    canActivate: [IsEnterpriseOrgGuard],
```
[the remark this answers is not in the document]
Comment (aj-rosado) [reply bitwarden_license/bit-web/src/app/tools/reports/reports-routing.module.ts]: Created a new task under this epic to follow this work

File: bitwarden_license/bit-web/src/app/tools/reports/member-access-report/member-access-report.component.html (hunk 1 above)
Comment (aj-rosado) [reply bitwarden_license/bit-web/src/app/tools/reports/member-access-report/member-access-report.component.html]: not necessary, removed

File: apps/web/src/app/admin-console/organizations/reporting/organization-reporting-routing.module.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (aj-rosado) [reply apps/web/src/app/admin-console/organizations/reporting/organization-reporting-routing.module.ts]: Thank you for the guidance!

I have followed @djsmith85 suggestion and removed the `reporting-route` and `module` on the `bitwarden-license` as the `member-access-report` is a standalone component and added it to the existing `organizations-routing.module`.

Removing all the imports from BL on the OSS

**Review by djsmith85 — APPROVED** [review]

File: bitwarden_license/bit-web/src/app/admin-console/organizations/organizations-routing.module.ts [hunk 4]
```diff
@@ -66,6 +67,23 @@ const routes: Routes = [
           },
         ],
       },
+      {
+        path: "reporting/reports",
+        canActivate: [AuthGuard],
```
Comment (eliykat) [inline bitwarden_license/bit-web/src/app/admin-console/organizations/organizations-routing.module.ts:72]: I think you probably want an `OrganizationPermissionsGuard` on this, probably referencing `org.canAccessReports`.

File: bitwarden_license/bit-web/src/app/tools/reports/member-access-report/member-access-report.component.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    protected reportService: MemberAccessReportService,
+    private searchService: SearchService,
+  ) {}
+
+  ngOnInit() {
+    this.dataSource.data = this.reportService.getMemberAccessMockData();
+
+    this._searchText$.pipe(
+      switchMap((searchText) => from(this.searchService.isSearchable(searchText))),
+      takeUntil(this.destroy$),
```
Comment (eliykat) [inline bitwarden_license/bit-web/src/app/tools/reports/member-access-report/member-access-report.component.ts:43]: This isn't doing anything - piping just returns a new observable, it's neither assigned anywhere nor subscribed to anywhere.

In general with the boilerplate search logic, it seems like you're following what's in other components, but (at least in AC's components), this is out of date and/or broken. A much simpler pattern can be seen in [new-base.people.component.ts](https://github.com/bitwarden/clients/blob/main/apps/web/src/app/admin-console/common/new-base.people.component.ts#L137-L140):

```ts
// connect the search input to a FormControl
protected searchControl = new FormControl("", { nonNullable: true });

constructor() {
    // Connect the search input to the table dataSource filter input
    this.searchControl.valueChanges
      .pipe(debounceTime(200), takeUntilDestroyed())
      .subscribe((v) => (this.dataSource.filter = v));
  }
```

File: bitwarden_license/bit-web/src/app/admin-console/organizations/organizations-routing.module.ts [hunk 6]
```diff
@@ -66,6 +67,23 @@ const routes: Routes = [
...
+        children: [
```
[the remark this answers is not in the document]
Comment (eliykat) [reply bitwarden_license/bit-web/src/app/admin-console/organizations/organizations-routing.module.ts:73]: Also, does this need to be feature flagged? It looks like a WIP.

**Review by eliykat — CHANGES_REQUESTED** [review]

File: bitwarden_license/bit-web/src/app/admin-console/organizations/organizations-routing.module.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (aj-rosado) [reply bitwarden_license/bit-web/src/app/admin-console/organizations/organizations-routing.module.ts]: The feature flag is on the `IsEnterpriseGuard` as it was added just for this.

[the remark this answers is not in the document]
Comment (eliykat) [reply bitwarden_license/bit-web/src/app/admin-console/organizations/organizations-routing.module.ts]: Sure - for next time, we have `FeatureFlagGuard` so you don't have to re-implement this functionality yourself.

File: libs/common/src/admin-console/abstractions/organization/organization.service.abstraction.ts [hunk 7]
```diff
@@ -82,6 +82,10 @@ export function canAccessImport(i18nService: I18nService) {
   );
 }
 
+export function canAccessReports(org: Organization): boolean {
+  return org.canAccessReports;
+}
```
Comment (eliykat) [inline libs/common/src/admin-console/abstractions/organization/organization.service.abstraction.ts:87]: I don't think there's any benefit to defining this new function when you can just access `org.canAccessReports` directly. The other functions here like `canAccessOrgAdmin` aggregate multiple permissions, but this doesn't.

So in the routing module:

```ts
data: { organizationPermissions: org => org.canAccessReports }
```

**Review by eliykat — CHANGES_REQUESTED** [review]
The main request here is re: `canAccessReports`, the other comments are non-blocking but just stuff I noticed.

**Review by eliykat — DISMISSED** [review]

**Review by djsmith85 — APPROVED** [review]

**Review by eliykat — CHANGES_REQUESTED** [review]
We just updated the `OrganizationPermissionsGuard` to use functional route guards: https://github.com/bitwarden/clients/pull/9631

This has been merged, which is probably causing your conflict. It also means you need to update your use of the route guard. It is now a function that takes the callback as a parameter, e.g. `OrganizationPermissionsGuard(org => org.canAccessReports)`. You no longer use the `data` property.

Sorry for the bad timing!

**Review by djsmith85 — APPROVED** [review]

**Review by eliykat — APPROVED** [review]
Thanks @aj-rosado !

**Pull request merged**